cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java

```diff
@@ -0,0 +1,203 @@
+
+  private LeaderElection leaderElection;
+  private ResourceCoordinator streamsResourceCoordinator;
+  private Discoverable handlerDiscoverable;
+  private Cancellable handlerSubscription;
+
+  @Inject
+  public DistributedStreamCoordinator(StreamAdmin streamAdmin, ZKClient zkClient,
+                                      DiscoveryServiceClient discoveryServiceClient,
+                                      StreamMetaStore streamMetaStore) {
+    super(streamAdmin);
+    this.zkClient = zkClient;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.streamMetaStore = streamMetaStore;
+    this.streamsResourceCoordinatorClient = new ResourceCoordinatorClient(zkClient);
+    this.handlerDiscoverable = null;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    Preconditions.checkNotNull(handlerDiscoverable, "Stream Handler discoverable has not been set");
+    streamsResourceCoordinatorClient.startAndWait();
+
+    // Start the resource coordinator that will map Streams to Stream handlers
+    leaderElection = new LeaderElection(
+      zkClient, "/election/" + STREAMS_COORDINATOR, new ElectionHandler() {
+      @Override
+      public void leader() {
+        LOG.info("Became Stream handler leader. Starting resource coordinator.");
+        streamsResourceCoordinator = new ResourceCoordinator(zkClient, discoveryServiceClient,
+                                                             new BalancedAssignmentStrategy());
+        streamsResourceCoordinator.startAndWait();
+
+        try {
+          // Create one requirement for the resource coordinator for all the streams.
+          // One stream is identified by one partition
+          ResourceRequirement.Builder builder = ResourceRequirement.builder(STREAMS_RESOURCE);
+          for (StreamSpecification spec : streamMetaStore.listStreams()) {
+            LOG.debug("Adding {} stream as a resource to the coordinator to manager streams leaders.", spec.getName());
+            builder.addPartition(new ResourceRequirement.Partition(spec.getName(), 1));
+          }
+          streamsResourceCoordinatorClient.submitRequirement(builder.build()).get();
+        } catch (Throwable e) {
+          LOG.error("Could not create requirement for coordinator in Stream handler leader", e);
+          Throwables.propagate(e);
+        }
+      }
+
+      @Override
+      public void follower() {
+        LOG.info("Became Stream handler follower.");
+        if (streamsResourceCoordinator != null) {
+          streamsResourceCoordinator.stopAndWait();
+        }
+      }
+    });
+
+    handlerSubscription = streamsResourceCoordinatorClient.subscribe(handlerDiscoverable.getName(),
+                                                                     new StreamsLeaderHandler());
+  }
+
+  @Override
+  protected void doShutDown() throws Exception {
+    // revoke subscription to resource coordinator
+    if (leaderElection != null) {
+      leaderElection.stopAndWait();
```

Comment:
Better wait with a timeout. `Uninterruptibles.getUninterruptibly(leaderElection.stop(), 5, TimeUnit.SECONDS);`

It's to avoid if the leader election failed to delete the node due to ZK issue, it won't block forever.